client/allocator/src/freeing_bump.rs

```diff
@@ -601,21 +657,72 @@ impl<'a> Drop for PoisonBomb<'a> {
 mod tests {
 	use super::*;
 
-	const PAGE_SIZE: u32 = 65536;
-
 	/// Makes a pointer out of the given address.
 	fn to_pointer(address: u32) -> Pointer<u8> {
 		Pointer::new(address)
 	}
 
+	#[derive(Debug)]
+	struct MemoryInstance {
+		data: Vec<u8>,
+		max_wasm_pages: u32,
+	}
+
+	impl MemoryInstance {
+		fn with_pages(pages: u32) -> Self {
+			Self { data: vec![0; (pages * PAGE_SIZE) as usize], max_wasm_pages: MAX_WASM_PAGES }
+		}
+
+		fn set_max_wasm_pages(&mut self, max_pages: u32) {
+			self.max_wasm_pages = max_pages;
+		}
+	}
+
+	impl Memory for MemoryInstance {
+		fn with_access<R>(&self, run: impl FnOnce(&[u8]) -> R) -> R {
+			run(&self.data)
+		}
+
+		fn with_access_mut<R>(&mut self, run: impl FnOnce(&mut [u8]) -> R) -> R {
+			run(&mut self.data)
+		}
+
+		fn pages(&self) -> u32 {
+			pages_from_size(self.data.len() as u64).unwrap()
+		}
+
+		fn max_pages(&self) -> Option<u32> {
+			Some(self.max_wasm_pages)
+		}
+
+		fn grow(&mut self, pages: u32) -> Result<(), ()> {
+			if self.pages() + pages > self.max_wasm_pages {
+				Err(())
+			} else {
+				self.data.resize(((self.pages() + pages) * PAGE_SIZE) as usize, 0);
+				Ok(())
+			}
+		}
+	}
+
+	#[test]
+	fn test_pages_from_size() {
+		assert_eq!(pages_from_size(0).unwrap(), 0);
+		assert_eq!(pages_from_size(1).unwrap(), 1);
+		assert_eq!(pages_from_size(65536).unwrap(), 1);
+		assert_eq!(pages_from_size(65536 + 1).unwrap(), 2);
+		assert_eq!(pages_from_size(2 * 65536).unwrap(), 2);
+		assert_eq!(pages_from_size(2 * 65536 + 1).unwrap(), 3);
+	}
+
 	#[test]
 	fn should_allocate_properly() {
 		// given
-		let mut mem = [0u8; PAGE_SIZE as usize];
+		let mut mem = MemoryInstance::with_pages(1);
 		let mut heap = FreeingBumpHeapAllocator::new(0);
```

Comment:
Yeah should probably now be achievable, before with just operating on the bytes it would have required a trait (which we now got). Maybe some possible future refactor, but is also not such a huge problem.

Reply:
In case of `wasmtime` the actual memory is owned by the executor, so to access it you also need a mutable handle to the context, but you're also storing the `FreeingBumpHeapAllocator` inside the context, and you can't have two mutable references to the same thing. That's why Basti has to store the allocator inside of an `Option` and temporarily `take()` it in `client/executor/wasmtime/src/host.rs`.

Technically you *could* add a field inside of the `FreeingBumpHeapAllocator` which would store a handle to the memory, but it still wouldn't own the memory (because it's owned by `wasmtime` and the `Memory` it exposes is just essentially an `usize` index into `wasmtime`'s internal store) and you still would need `wasmtime`'s context to access it. So there isn't really much point in doing that.